Sources/NIOHPACK/SimpleRingBuffer.swift

```diff
@@ -0,0 +1,609 @@
+            // wrap it around
+            newIndex -= self._storage.capacity
+            // now test it
+            assert(newIndex <= self._ringHead)
+        }
+        
+        self._moveTail(to: newIndex)
+        self._readableBytes += amount
+    }
+    
+    @inlinable
+    mutating func moveHead(forwardBy amount: Int) {
+        self._moveHead(forwardBy: amount)
+    }
+    
+    @inlinable
+    mutating func moveTail(forwardBy amount: Int) {
+        self._moveTail(forwardBy: amount)
+    }
+    
+    @inlinable
+    mutating func unwrite(byteCount: Int) {
+        assert(byteCount >= 0)
+        if byteCount <= self._ringTail {
+            self._ringTail -= byteCount
+        } else {
+            self._ringTail = self._storage.capacity - byteCount + self._ringTail
+        }
+        self._readableBytes -= byteCount
+    }
+    
+    @inlinable
+    mutating func unread(byteCount: Int) {
+        assert(byteCount >= 0)
+        if byteCount <= self._ringHead {
+            self._ringHead -= byteCount
+        } else {
+            self._ringHead = self._storage.capacity - byteCount + self._ringHead
+        }
+        self._readableBytes += byteCount
+    }
+    
+    @inlinable
+    mutating func _set<S: ContiguousCollection>(bytes: S, at index: Int) throws -> Int where S.Element == UInt8 {
+        let totalCount = bytes.count
+        guard totalCount <= self.writableBytes else {
+            throw RingBufferError.BufferOverrun(amount: Int(totalCount - self._storage.capacity))
+        }
+        
+        let firstPart = self._storage.capacity - index
+        if firstPart >= bytes.count {
+            _storage.set(bytes: bytes, at: index)
+        } else {
+            let secondPart = bytes.count - firstPart
+            bytes.withUnsafeBytes { ptr in
+                _storage.set(bytes: UnsafeRawBufferPointer(start: ptr.baseAddress!, count: firstPart), at: index)
+                _storage.set(bytes: UnsafeRawBufferPointer(start: ptr.baseAddress!.advanced(by: firstPart), count: secondPart), at: 0)
+            }
+        }
+        return bytes.count
+    }
+    
+    @inlinable
+    mutating func _set<S : Sequence>(bytes: S, at index: Int) throws -> Int where S.Element == UInt8 {
+        assert(!([Array<S.Element>.self, StaticString.self, ContiguousArray<S.Element>.self, UnsafeRawBufferPointer.self, UnsafeBufferPointer<UInt8>.self].contains(where: { (t: Any.Type) -> Bool in t == type(of: bytes) })),
+               "called the slower set<S: Sequence> function even though \(S.self) is a ContiguousCollection")
+        guard bytes.underestimatedCount <= self.writableBytes else {
+            throw RingBufferError.BufferOverrun(amount: Int(self._storage.capacity) - bytes.underestimatedCount)
+        }
+        
+        // it *should* fit. Let's see though.
+        let underestimatedCount = bytes.underestimatedCount
+        
+        return try self._storage.withVeryUnsafeBytes { ptr in
```

Comment:
This all seems more complex than it needs to be. Presumably we are going to do at most two copies. In that case, 
why not split the `Sequence` and do two smaller sets using `Sequence.prefix()` and `Sequence.dropFirst()`? Those sets can be done on the underlying buffer. In fact, you can use `Sequence.dropFirst().prefix()` to avoid over-copying, and then confirm that there are no bytes left over in the sequence after that if you want to avoid the over-copy issue.

Reply:
I'd been leery of this because I didn't want to automatically incur the overhead of iterating through a potentially large sequence just to get the count if I might not require it. Given that we're now only using this for types which at least conform to `Collection`, I think I will just change this to use `Collection` rather than `Sequence`, and make use of the concrete `count` property.

Reply:
That seems reasonable enough to me.